addons/zylann.hterrain/native/src/quad_tree_lod.cpp

```diff
@@ -0,0 +1,248 @@
+	int po = 0;
+	while (full_size > base_size) {
+		full_size = full_size >> 1;
+		po += 1;
+	}
+	return po;
+}
+
+// The higher, the longer LODs will spread and higher the quality.
+// The lower, the shorter LODs will spread and lower the quality.
+void QuadTreeLod::set_split_scale(real_t p_split_scale) {
+	real_t MIN = 2.0f;
+	real_t MAX = 5.0f;
+
+	// Split scale must be greater than a threshold,
+	// otherwise lods will decimate too fast and it will look messy
+	if (p_split_scale < MIN)
+		p_split_scale = MIN;
+	if (p_split_scale > MAX)
+		p_split_scale = MAX;
+
+	_split_scale = p_split_scale;
+}
+
+real_t QuadTreeLod::get_split_scale() {
+	return _split_scale;
+}
+
+// Intention is to only clear references to children
+void QuadTreeLod::_clear_children(unsigned int index) {
+    Quad* quad = _get_node(index);
+    if (quad->has_children()) {
+        _recycle_children(quad->first_child);
+        quad->first_child = NO_CHILDREN;
+    }
+}
+
+// Returns the index of the first_child. Allocates from _free_indices.
+unsigned int QuadTreeLod::_allocate_children() {
+    if (_free_indices.size() == 0) {
+        return NO_CHILDREN;
+    }
+
+    unsigned int i0 = _free_indices[_free_indices.size() - 1];
+    _free_indices.pop_back();
+    return i0;
+}
+
+// Pass the first_child index, not the parent index. Stores back in _free_indices.
+void QuadTreeLod::_recycle_children(unsigned int i0) {
+    // Debug check, there is no use case in recycling a node which is not a first child
+    CRASH_COND(i0 % 4 != 0);
+
+    for (int i = 0; i < 4; ++i) {
+        _node_pool[i0 + i].init();
+    }
+
+    _free_indices.push_back(i0);
+}
+
+Variant QuadTreeLod::_make_chunk(int lod, int origin_x, int origin_y) {
+    if (_make_func.is_valid()) {
+        return _make_func->call_func(origin_x, origin_y, lod);
+    } else {
+        return Variant();
+    }
+}
+
+void QuadTreeLod::_recycle_chunk(unsigned int quad_index, int lod) {
+    Quad* quad = _get_node(quad_index);
+    if (_recycle_func.is_valid()) {
+        _recycle_func->call_func(quad->get_data(), quad->origin_x, quad->origin_y, lod);
+    }
+}
+
+void QuadTreeLod::clear() {
+	_join_all_recursively(ROOT, _max_depth);
+	_max_depth = 0;
+	_base_size = 0;
+}
+
+void QuadTreeLod::_join_all_recursively(unsigned int quad_index, int lod) {
+    Quad* quad = _get_node(quad_index);
+
+    if (quad->has_children()) {
+        for (int i = 0; i < 4; i++) {
+            _join_all_recursively(quad->first_child + i, lod - 1);
+        }
+        _clear_children(quad_index);
+
+    } else if (quad->is_valid()) {
+        _recycle_chunk(quad_index, lod);
+        quad->clear_data();
+    }
+}
+
+void QuadTreeLod::create_from_sizes(int base_size, int full_size) {
+    clear();
+    _base_size = base_size;
+    _max_depth = compute_lod_count(base_size, full_size);
+
+    int node_count = static_cast<int>(pow(4, _max_depth) / (4 - 1)) - 1;    // # nodes is N^L / (N - 1). -1 for root, where N=num children, L=levels
```

Comment:
We should be lucky `pow` uses `double`, because `float` might be unable to accurately contain some of the biggest results this can get. Or we could try using an integer version

Reply:
Thinking back, the maximum size in the plugin at the moment is 4096, and minimum chunk size is 16. Considering `Quad` is `(32+4+4+4)` bytes, that means the maximum amount of nodes we can have is:
```
>>> (32+4+4+4) * (int(pow(4, 9) / (4 - 1)) - 1)
3844720
```
3.8 megabytes is quite a lot for a preallocation strategy. It also will never get reached, unless you bump LOD distance such that the entire terrain is visible at LOD0.
Higher chunk size (so less quads) is still recommended, but maybe in the future we'll switch away from preallocating, depending on design changes.

Reply:
I don't see any change requests here, so I won't make any for now.